### PR TITLE
Data in Modules, Modern facts & Cleanup

### LINKED ISSUE #304: Do not fail fatal if OS is unsupported.
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 6.0.2
- Ruby:
- Distribution: Ubuntu 18.04
- Module version: 7.3.0

## Any additional information you'd like to impart

The module works totally fine under Ubuntu 18.04. I had no issues but I also do not test some edge cases.

Anyway please do not fail if the version is unsupported. I can understand if you auto config may not work for unsupported version but let the user the chance to configure the module manually.

### PR DESCRIPTION
**Pull request by jkroepke**
PR is still in progress.

Reviews are welcome.

#### Pull Request (PR) description
* Add Data in modules
* Add defaults for unsupported OS
* Used modern facts
* Removed top scope variables
* Removed create_resources
* Removed params.pp
* Remove systemd option in favor of facts provided by stdlib

#### This Pull Request (PR) fixes the following issues
Fixes #304 

#### Tested
* Ubuntu 18.04 (no fresh installation, no changes)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM ruby:2.4
```
Comment (bastelfreak) [inline Dockerfile]: I'm not sure if we want a Dockerfile. If so, we should probably contribute it to https://github.com/voxpupuli/modulesync_config/tree/master/moduleroot

File: manifests/deploy/prepare.pp [hunk 2]
```diff
@@ -19,9 +21,9 @@
 # Phil Bayfield https://bitbucket.org/Philio/
 #
 
-class openvpn::deploy::prepare {
-
-  class { 'openvpn::params': }
+class openvpn::deploy::prepare(
+  String $etc_directory,
```
Comment (bastelfreak) [inline manifests/deploy/prepare.pp]: We can validate this even stricter:
```suggestion
  Stdlib::Absolutepath $etc_directory,
```

File: manifests/init.pp [hunk 3]
```diff
@@ -69,35 +96,82 @@
...
-  Hash $clients                         = {},
-  Hash $client_specific_config_defaults = {},
-  Hash $client_specific_configs         = {},
-  Hash $revoke_defaults                 = {},
-  Hash $revokes                         = {},
-  Hash $server_defaults                 = {},
-  Hash $servers                         = {},
+  Boolean                    $autostart_all,
+  Boolean                    $manage_service,
+  String                     $etc_directory,
```
Comment (bastelfreak) [inline manifests/init.pp]:
```suggestion
  Stdlib::Absolutepath       $etc_directory,
```

File: manifests/init.pp [hunk 4]
```diff
@@ -69,35 +96,82 @@
...
+  String                     $group,
```
Comment (bastelfreak) [inline manifests/init.pp]:
```suggestion
  String[1]                  $group,
```

File: manifests/init.pp [hunk 5]
```diff
@@ -69,35 +96,82 @@
...
+  Boolean                    $link_openssl_cnf,
+  Optional[Stdlib::Unixpath] $pam_module_path,
+  Boolean                    $namespecific_rclink,
+  Pattern[/^[23]\.0$/]       $default_easyrsa_ver,
+  Stdlib::Unixpath           $easyrsa_source,
+  Variant[String, Array]     $additional_packages,
```
Comment (bastelfreak) [inline manifests/init.pp]:
```suggestion
  Variant[String[1], Array[String[1]]] $additional_packages,
```

File: manifests/install.pp [hunk 6]
```diff
@@ -31,15 +31,17 @@
...
 #
-class openvpn::install inherits openvpn::params {
+class openvpn::install {
+
+  include openvpn
 
   ensure_packages(['openvpn'])
-  if $::openvpn::params::additional_packages != undef {
-    ensure_packages( any2array($::openvpn::params::additional_packages) )
+  if $openvpn::additional_packages != undef {
```
Comment (bastelfreak) [inline manifests/install.pp]: This could be simplified to:
```suggestion
  if $openvpn::additional_packages {
```

File: data/defaults.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+openvpn::etc_directory: '/etc'
+openvpn::group: 'nobody'
+openvpn::link_openssl_cnf: true
+openvpn::pam_module_path: ~
+openvpn::namespecific_rclink: false
+openvpn::default_easyrsa_ver: '2.0'
+openvpn::easyrsa_source: '/usr/share/easy-rsa/'
+openvpn::additional_packages: ['easy-rsa']
+openvpn::ldap_auth_plugin_location: ~
+openvpn::systemd: false
```
Comment (bastelfreak) [inline data/defaults.yaml]: Do we need this parameter?
```
$ git grep systemd
CHANGELOG.md:* Add systemd support for Debian >= 8.0 ([#161](https://github.com/luxflux/puppet-openvpn/pull/161))
CHANGELOG.md:* Support for systemd ([#127](https://github.com/luxflux/puppet-openvpn/pull/127))
HISTORY.md:* Add systemd support for Debian >= 8.0 ([#161](https://github.com/luxflux/puppet-openvpn/pull/161))
HISTORY.md:* Support for systemd ([#127](https://github.com/luxflux/puppet-openvpn/pull/127))
data/defaults.yaml:openvpn::systemd: false
manifests/init.pp:  if $facts['service_provider'] != 'systemd' {
manifests/server.pp:  if $facts['service_provider'] == 'systemd' and $openvpn::namespecific_rclink {
manifests/server.pp:    fail("Using systemd and namespecific rclink's (BSD-style) is not allowed")
manifests/server.pp:    if $facts['service_provider'] == 'systemd' {
manifests/server.pp:  if $facts['service_provider'] == 'systemd' {
manifests/server.pp:        provider => 'systemd',
spec/classes/openvpn_init_spec.rb:        context 'system without systemd' do
spec/classes/openvpn_init_spec.rb:            service_provider: 'systemd'
spec/classes/openvpn_init_spec.rb:        context 'system with systemd' do
```

**Comment by bastelfreak** [comment]
Fixes https://github.com/voxpupuli/puppet-openvpn/issues/306

**Comment by bastelfreak** [comment]
Thanks for the awesome PR @jkroepke. I only did some little inline comments, but I didn't spot anything bad. Can you also add Ubuntu 18.04 to our acceptance test matrix? That would be:
```diff
diff --git a/.sync.yml b/.sync.yml
index d485814..e241dff 100644
--- a/.sync.yml
+++ b/.sync.yml
@@ -1,6 +1,7 @@
 ---
 .travis.yml:
   docker_sets:
+    - set: ubuntu1804-64
     - set: ubuntu1604-64
     - set: ubuntu1404-64
     - set: centos7-64
diff --git a/.travis.yml b/.travis.yml
index 1b0d04c..3566dc8 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -25,6 +25,24 @@ matrix:
   - rvm: 2.4.4
     bundler_args: --without system_tests development release
     env: PUPPET_VERSION="~> 5.0" CHECK=build DEPLOY_TO_FORGE=yes
+  - rvm: 2.5.1
+    bundler_args: --without development release
+    dist: trusty
+    env: PUPPET_INSTALL_TYPE=agent BEAKER_IS_PE=no BEAKER_PUPPET_COLLECTION=puppet5 BEAKER_debug=true BEAKER_setfile=ubuntu1804-64{hypervisor=docker} CHECK=beaker
+    services: docker
+    sudo: required
+  - rvm: 2.5.1
+    bundler_args: --without development release
+    dist: trusty
+    env: PUPPET_INSTALL_TYPE=agent BEAKER_IS_PE=no BEAKER_PUPPET_COLLECTION=puppet6 BEAKER_debug=true BEAKER_setfile=ubuntu1804-64{hypervisor=docker} CHECK=beaker
+    services: docker
+    sudo: required
+  - rvm: 2.5.1
+    bundler_args: --without development release
+    dist: trusty
+    env: PUPPET_INSTALL_TYPE=agent BEAKER_IS_PE=no BEAKER_PUPPET_COLLECTION=puppet6-nightly BEAKER_debug=true BEAKER_setfile=ubuntu1804-64{hypervisor=docker} CHECK=beaker
+    services: docker
+    sudo: required
   - rvm: 2.5.1
     bundler_args: --without development release
     dist: trusty
```

If that doesn't work we can still implement it in another PR, to keep the diff small in this one.

**Comment by jkroepke** [comment]
Thanks for the reviews!

@bastelfreak about the acceptance test. I would add it but I can not find a nodeset for 1804 here. Is this correct?

https://github.com/voxpupuli/puppet-openvpn/tree/master/spec/acceptance/nodesets

**Comment by bastelfreak** [comment]
The nodesets for docker are generated on the fly. It's triggered by `BEAKER_setfile=ubuntu1804-64{hypervisor=docker}` which utilities https://github.com/puppetlabs/beaker-hostgenerator#beaker-host-generator.

File: manifests/config.pp [hunk 8]
```diff
@@ -29,12 +29,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-class openvpn::config inherits openvpn::params {
+class openvpn::config {
 
-  if $::osfamily == 'Debian' {
+  if $facts['osfamily'] == 'Debian' {
```
Comment (jkroepke) [inline manifests/config.pp]: @bastelfreak I have to use `$facts['osfamily']` here. `$facts['os']['family']` works inside puppet fine but i get `[] Operator on Undef is not allowed` on rspec tests.

Any idea?

File: manifests/client.pp [hunk 9]
```diff
@@ -354,7 +354,7 @@
 
   file { "${etc_directory}/openvpn/${server}/download-configs/${name}/${name}.conf":
     owner   => root,
-    group   => $::openvpn::params::root_group,
+    group   => 0,
```
Comment (bastelfreak) [inline manifests/client.pp]: Why did you switch away from the parameter to a hardcoded number? I don't think that there are a lot of usecases to set this to something else, but since the parameter is already there we could keep it?

File: manifests/ca.pp [hunk 10]
```diff
@@ -150,7 +150,7 @@
         require => File["${etc_directory}/openvpn/${name}/easy-rsa"],
       }
 
-      if $openvpn::params::link_openssl_cnf == true {
+      if $openvpn::link_openssl_cnf == true {
```
Comment (bastelfreak) [inline manifests/ca.pp]: nitpick: if you want to clean it up, you could do:
```suggestion
      if $openvpn::link_openssl_cnf {
```

But that is really not required

File: manifests/client.pp (hunk 9 above)
Comment (jkroepke) [reply manifests/client.pp]: I would prefer to keep it, because the root group is always 0.

But its `backwards-incompatible`, expect you want this parameter.

File: manifests/ca.pp [hunk 11]
```diff
@@ -216,7 +216,7 @@
...
```
Comment (bastelfreak) [inline manifests/ca.pp]:
```suggestion
      if $openvpn::link_openssl_cnf  {
```

File: manifests/init.pp [hunk 12]
```diff
@@ -69,35 +96,82 @@
...
+  -> Class['openvpn::config']
   -> Class['openvpn']
 
-  if !$::openvpn::params::systemd {
+  if $facts['service_provider'] != 'systemd' {
     class { 'openvpn::service':
       subscribe => [Class['openvpn::config'], Class['openvpn::install'] ],
     }
+
     if empty($servers) {
```
Comment (jkroepke) [inline manifests/init.pp]: @bastelfreak What about this?

It can be just `if !$servers` or  `if $servers != {}`

File: manifests/config.pp (hunk 8 above)
Comment (bastelfreak) [reply manifests/config.pp]: I assume because of this:
https://github.com/voxpupuli/puppet-openvpn/blob/139c72778008b91469d445f4f2c89d5c35d57ef4/spec/unit/openvpn_module_spec.rb#L15-L56

It explicitly mocks the `osfamily` hash and doesn't use rspec-puppet-facts.

File: manifests/init.pp (hunk 12 above)
Comment (bastelfreak) [reply manifests/init.pp]: mhm, good question. We don't have anything for this in our styleguide. On a first guess `empty($servers)` looks as good as `if !$servers`.

**Comment by bastelfreak** [comment]
I am fine with this as is, but would like to see another review

File: manifests/config.pp (hunk 8 above)
Comment (Dan33l) [reply manifests/config.pp]: structured facts with rspec are facts[:os]['family'].
hope this help.

**Comment by Dan33l** [comment]
Interesting PR, thank you.
I didn't saw weird things and i am fine also.
